Flux Elliptic Problem 2

Hunti0 opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [x] Example
- [ ] Documentation

## Description
Implementation of flux operator and usage of MOLE library to solve Problem 2 found in "Mimetic Discretization of Elliptic PDE with Full Tensor Coefficients" - Huy Vu and Jose Castillo; November 2006

## Added/updated tests?
- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [x] I need help with writing tests

## Read Contributing Guide and Code of Conduct

- [x] I have read and followed the [Contributing Guide](https://github.com/csrc-sdsu/mole/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/csrc-sdsu/mole/blob/main/CODE_OF_CONDUCT.md)

## [optional] Are there any post deployment tasks we need to perform?
No
## [optional] What gif best describes this PR or how it makes you feel?

<img src="https://miro.medium.com/v2/1*VpQb2kbPdj6vXH2pJqi7Qg.gif" width="100" height="100" />

Hunti0 committed:
Update EllipticFlux2D Problem 2

aboada commented:
Need to see this PR from my PC, but shouldn't it be from your branch to your repo's master?

See attached 

![Screenshot_20250813-174643.png](https://github.com/user-attachments/assets/e1fac332-60c5-4b63-9b52-33539fe955ae)

aboada commented on examples/cpp/EllipticFlux2D Problem 2:
```diff
@@ -1,42 +1,39 @@
 #include "mole.h"
 #include <vector>
 
 using namespace arma;
```
I don't recommend using a complete namespace. See: 
https://google.github.io/styleguide/cppguide.html#Namespaces

aboada commented on examples/cpp/EllipticFlux2D Problem 2:
```diff
@@ -1,42 +1,39 @@
 
-// Analytic solution
 double exact_u(double x, double y) {
     const double lambda = 1.0;
```
nit: Can we declare this `lambda` as `constexpr` and define it as a global variable within this file?

aboada commented on examples/cpp/EllipticFlux2D Problem 2:
```diff
@@ -1,42 +1,39 @@
 
 int main() {
-    std::vector<u32> Ns = {10, 20, 40, 80}; 
+    std::vector<u32> Ns = {10, 20, 40, 80};
```
What do these numbers mean (`10, 20, 40, 80`)? What is `Ns`?

aboada commented on examples/cpp/EllipticFlux2D Problem 2:
```diff
@@ -1,42 +1,39 @@
 
     for (size_t ni = 0; ni < Ns.size(); ++ni) {
         u32 m = Ns[ni], n = Ns[ni];
         double Lx = 1.0, Ly = 1.0;
```
If these variables (`Lx`, `Ly`, etc.) are not expected to change, declare them as constants, potentially as `constexpr`. Define them outside of the loop.

aboada commented on examples/cpp/EllipticFlux2D Problem 2:
```diff
@@ -1,42 +1,39 @@
 
-        // Build analytic solution and RHS on grid 
+        // Build analytic solution and RHS on grid
         vec uex((m+2)*(n+2)), f((m+2)*(n+2));
```
What is `uex`? Any opportunities to improve the naming of the vector to improve readability? A comment can also help.

aboada commented on examples/cpp/EllipticFlux2D Problem 2:
```diff
@@ -1,42 +1,39 @@
+        // Build analytic solution and RHS on grid
         vec uex((m+2)*(n+2)), f((m+2)*(n+2));
         for (u32 i = 0; i < m + 2; ++i)
             for (u32 j = 0; j < n + 2; ++j) {
```
Do we need this calculation?
```
u32 idx = i * (n + 2) + j;
```
What happens if we just start an index from 0 outside of the loop for the `i` index and we increment it inside the `j` loop?

aboada commented on examples/cpp/EllipticFlux2D Problem 2:
```diff
@@ -45,27 +42,32 @@ int main() {
 
+        // Assemble Laplacian operator
         Laplacian L(k, m, n, dx, dy);
         sp_mat Amat = static_cast<sp_mat>(L);
```
If I remember correctly, `Laplacian` extends from `sp_mat`. Why do we need to do a `static_cast` to `sp_mat` here?

Hunti0 replied:
I static cast because I tend to get an undefined operator error when interacting with the MOLE operators and that tends to fix the issue

aboada commented on examples/cpp/EllipticFlux2D Problem 2:
```diff
@@ -45,27 +42,32 @@ int main() {
+        // Robin boundary condition parameters
         const double alpha = -std::exp(lambda);
         const double beta  = (std::exp(lambda) - 1.0) / lambda;
         sp_mat BCmat = static_cast<sp_mat>(
```
same as the comment about the `Laplacian`.

aboada commented on examples/cpp/EllipticFlux2D Problem 2:
```diff
@@ -45,27 +42,32 @@ int main() {
-        // Compute Mean Square Error
-        double mean_square = norm(sol - uex, 2) * std::sqrt(dx * dy);
+        // Compute global L2 error
+        double err = norm(sol - uex, 2) / norm(uex, 2);
```
So what kind of values are we getting here?

aboada commented on examples/cpp/EllipticFlux2D Problem 2:
```diff
@@ -1,42 +1,39 @@
 #include <vector>
 
 using namespace arma;
 using std::cout;
```
My personal preference is just to reference `std::cout` directly but I leave it up to you.

aboada reviewed (changes requested):
Thanks for the PR, Hunter!

It would be good if we could see the actual changes on this PR. It seems your `master` already has some commits and does not coincide with `Mole`. I recommend synching with the `Mole master` branch (as in, make your `master == Mole master`) and then add all relevant changes to this PR: the `README`, the `cmake` changes to allow compilation of your code, etc.

Hunti0 committed:
Added recommended changes

Made variables easier to understand, added a few more comments, changed indexing to start at 0

Hunti0 committed:
Merge branch 'test-branch' into master
